### PR TITLE
"Reveal in Tree"

### PR DESCRIPTION
**Pull request by arthur801031**
Associated Issue: #1290 #1370

### Summary of Changes

* Add code in `SourcesTree.js` to determine which item(s) should be expanded and pass the list to `ManagedTree.js`.
* Add code in `ManagedTree.js` to process the list of items that should be expanded.

### CODE REVIEW

File: src/components/SourcesTree.js [hunk 1]
```diff
@@ -38,6 +58,32 @@ let SourcesTree = React.createClass({
...
+      const itemsStrings = returnItemsStrings(sourceURL);
+      const processIndex = 0;
+
+      // Determine which item(s) should be added to itemsList
+      while (processIndex < itemsStrings.length) {
+        const i = 0;
+        const found = false;
+        while (!found) {
+          if (itemsStrings[processIndex] == sourceTreeList.contents[i].path) {
+            addToItemsList(sourceTreeList.contents[i], true);
```
Comment (jasonLaster) [inline src/components/SourcesTree.js]: hmm, I think you can use react `setState` here instead of redux.

Also, maybe set the items at the end as opposed to one at a time

File: src/components/SourceTabs.js [hunk 2]
```diff
@@ -114,13 +116,27 @@ const SourceTabs = React.createClass({
       click: () => closeTabs(tabs)
     };
 
-    showMenu(e, buildMenu([
+    const showSourceMenuItem = {
```
Comment (jasonLaster) [inline src/components/SourceTabs.js]: I landed this work earlier today (I think). A rebase would be healthy either way :)

**Comment by arthur801031** [comment]
Hi @jasonLaster

I have resolved all the previous issues. Please check!
I have finished the following tasks:
- [x] Rebase
- [x] Resolve failing test
- [x] Remove the list of items from redux
- [x] Use `setState` instead of redux

File: src/components/SourcesTree.js [hunk 3]
```diff
@@ -17,6 +17,19 @@ const actions = require("../actions");
 const Svg = require("./utils/Svg");
 const { throttle } = require("../utils/utils");
 
+function returnItemsStrings(url) {
+  url = url.replace("http:/", "");
+  url = url.replace("https:/", "");
```
Comment (jasonLaster) [inline src/components/SourcesTree.js]: perhaps  `url.replace(/http(s)?:\//, "");`

File: src/components/SourcesTree.js [hunk 4]
```diff
@@ -17,6 +17,19 @@ const actions = require("../actions");
...
+  const itemsStrings = [];
+  for (const i = 1; i < url.length; i++) {
+    if (url[i] == "/") {
+      itemsStrings.push(url.substring(0, i));
+    }
+  }
```
Comment (jasonLaster) [inline src/components/SourcesTree.js]: could we do the same thing `with url.split("/")` :)

File: src/components/SourcesTree.js [hunk 5]
```diff
@@ -44,6 +57,33 @@ let SourcesTree = React.createClass({
...
+        while (!found) {
+          if (itemsStrings[processIndex] == sourceTreeList.contents[i].path) {
+            tempList.push(sourceTreeList.contents[i]);
+            sourceTreeList = sourceTreeList.contents[i];
+            found = true;
+            processIndex++;
+          }
+          i++;
+        }
+      }
```
Comment (jasonLaster) [inline src/components/SourcesTree.js]: this is a little confusing to me... i'll play with it later

**Comment by arthur801031** [comment]
Hi @jasonLaster 

No worries :) 
I've made the changes that you suggested. Please check!

**Comment by jasonLaster** [comment]
@Dalimil, @bomsy, @gregtatum, @jlongster mind taking a look at this PR?

I think we can simplify the logic for expanding the tree, buy you guys have more experience than I do.

File: src/components/SourcesTree.js (hunk 5 above)
Comment (Dalimil) [inline src/components/SourcesTree.js]: I don't think that using while loops is a good idea - it's difficult to follow the indices and it might as well run forever.

I've never run this code but the `const` reassignments look very suspicious...

**Comment by Dalimil** [comment]
I don't think the code works... Constant variable reassignments will produce TypeErrors I think...

**Comment by arthur801031** [comment]
Hi @jasonLaster

I've been testing my code on http://cnn.com, and I found some cases which my previous code would crash. They have now been resolved with the exception of `extensions://`. This requires writing a special case. Please check and let me know if I need to handle this case. Thank you! :)

**Comment by jasonLaster** [comment]
CNN is a great test case. 

I do think we should handle extensions:// and I suppose javascript:// 

This would be great example of where we can write don't unit tests to build confidence. I'll be at my desk tomorrow and will help you stay them. I'm also optimistic about simplifying the code. But tests will help there:)

**Comment by arthur801031** [comment]
Hi @jasonLaster 

- I have added the code to handle `extensions://`, but I haven't come across a folder named `javascript://`, so I haven't implemented that.
- I have tested my code on http://cnn.com, https://www.cnet.com/, http://www.wsj.com/, and every file that I tested could be revealed in the tree.
- While I was testing the function on those websites, I saw errors in the console, so I commented out the code that I've written, and the program still produced those errors when I click on any folders and files.
screen shot 2016-12-12 at 2 01 19 pm

- Please let me know where unit tests should be placed and how to write them. I'm not really familiar with them. Thank you very much! :)

**Comment by jasonLaster** [comment]
Thanks @arthur801031 for the extensive testing. I think you'll like the testing aspect, it's our way of quickly working on a complicated feature like reveal and making sure that we can test all of the edge cases and then make changes quickly.

* I took the liberty of moving `getExpandedItems` to our `utils/sources-tree` module where we keep the utility functions for the tree.
* I added one test that you can run with either `yarn run test` or `yarn run mocha-server` and then see in the browser*. You should add lots more tests
* While working on the test I thought of a **MUCH** simpler approach to `getExpandedParts`. The way I understand your algorithm works, is that it tries to find each part of the path as you traverse down the tree. Here's what I propose: lets update the tree so that: a) it's easy to get the leaf nodes (sources), b) each node has a parent reference so you can get the ancestor directory. This way, we can re-write our function as a) find the source in the tree with the same path and then b) get it's ancestor directories.


[1] You'll need to apply this [gist](https://gist.github.com/jasonLaster/da3f9b855a732c5a42abfa2b6d10b5df) to see the tests in the browser, the runner sadly needs some love. You can also change your test to `it.only` to only run that test.

**Comment by jasonLaster** [comment]
Alright, so i messed around for a bit and this is what I learned. 

* The tree has a really helpful thing called a parentMap. the parentMap will help us find the a parent of any node, given the node
* the parentMap uses the tree nodes as keys, so we have to add a new `findSource` function for finding leaves
* we can get the directory path by using the parentMap
* this should be very similar to what expandedPaths did, but there might be edge cases! needs lots of testing :)

I pushed a bunch of new code that probably doesn't work :) didn't run it.

**Comment by arthur801031** [comment]
Hi @jasonLaster ,

I took a look at the `getDirectories ` method. From what I understand, `findSource` finds the leaf/source node by going into each directory and traverse through that directory's folders and files before moving onto the next directory. For example, if our target file is file2, `findSource` would check dir1 -> subdir1 -> subdir2 -> file1 -> dir2 -> subdir1 -> file2 to find the target file. Then, `getDirectories ` method would use that target file's leaf node to find its ancestors, which means processing file2 -> subdir1 -> dir2.
```
root
  |
  - - ->dir1
             |
             --->subdir1
             --->subdir2
                        |
                        ---> file1
  - - ->dir2
             |
             --->subdir1
                        |
                        ---> file2
```

`getExpandedItems`, on the other hand, traverse dir1 -> dir2 -> subdir1 -> file2, and while it finds dir2, subdir1, and file2, the program adds them to the array. So, I think `getExpandedItems` is faster and doesn't check other irrelevant folders and files and does not traverse back to find the target leaf's ancestor items.

Please let me know if this is correct. Thank you! :)

**Comment by jasonLaster** [comment]
Hi, @arthur801031 that's a good point. I don't think it should be a performance problem though, and if it is, we can always optimize it by maintaining a list of references to the leaf nodes.

**Comment by jlongster** [comment]
This does seem like a lot of complicated code so it'd take me a while to think about how I'd implement it to give reasonable feedback, so I don't have much to say right now. I'm not sure why in `componentWillReceiveProps` of `SourcesTree` you can't do a search of the entire tree for a specific source item and build a list of nested sources to expand when you find the right source. You wouldn't have to add anything to `utils/source-tree.js` then; you already have the source tree structure, so just search it?

**Comment by arthur801031** [comment]
I just finished `findSource ` and `getDirectories`, and I've tested on http://cnn.com. It seems to be working without any issue. Please check! Thank you!

I tried to git squash my commits, but after I saved the file, I received this error:
```
error: could not apply 73add1d... Add show source redux backend (#1370)
When you have resolved this problem, run "git rebase --continue".
```
Do I need to manually resolve the conflicts?

**Comment by arthur801031** [comment]
Hi @jasonLaster ,

I have cleaned up the code and added 1 additional unit test for reveal. I have tested the code on  http://cnn.com and https://www.cnet.com/. If you would like the code to be simpler, you could remove the `else if ` from line 280 ~ line 287 in `findSource`. This special case only came up 3 times when I tested the code on those 2 websites. Please check. Thank you! :)

special2

special2-cnet

special2-cnet2

**Comment by jasonLaster** [comment]
Thanks @arthur801031! I just squashed your work which was crazy in its own right :)

Will review now some of the other details

File: src/components/utils/ManagedTree.js [hunk 6]
```diff
@@ -10,10 +10,29 @@ let ManagedTree = React.createClass({
...
       expanded: new Set(),
-      focusedItem: null
+      focusedItem: null,
+      listItems: []
     };
   },
 
+  componentWillReceiveProps(nextProps) {
+    this.setState({ listItems: nextProps.listItems });
+  },
```
Comment (jasonLaster) [inline src/components/utils/ManagedTree.js]: do we need state here? I think we can rely on just using `props.listItems`

File: src/components/utils/ManagedTree.js [hunk 7]
```diff
@@ -10,10 +10,29 @@ let ManagedTree = React.createClass({
...
+
+  componentDidUpdate(nextProps) {
+    const listItemsLocal = this.state.listItems;
+    if (listItemsLocal && listItemsLocal.length > 0) {
```
Comment (jasonLaster) [inline src/components/utils/ManagedTree.js]: if we don't have a listItemsLocal , lets bail

File: src/components/utils/ManagedTree.js [hunk 8]
```diff
@@ -10,10 +10,29 @@ let ManagedTree = React.createClass({
...
+      if (listItemsLocal.length == 1) {
+        this.focusItem(listItemsLocal[0]);
+      }
+      else {
+        this.expandListItems(listItemsLocal);
+      }
+      listItemsLocal.splice(listItemsLocal.length - 1, 1);
+      this.setState({ listItems: listItemsLocal });
```
Comment (jasonLaster) [inline src/components/utils/ManagedTree.js]: this seems weird to me... perhaps we do need some state, but the name collision makes it hard to follow...

Also, if we have an expanded list, do we need this listItemsLocal list... and if it does to an empty list because it keeps getting spliced down, can we do this another way. I guess i'm saying, this seems weird :)

File: src/components/utils/ManagedTree.js [hunk 9]
```diff
@@ -31,6 +50,13 @@ let ManagedTree = React.createClass({
     }
   },
 
+  expandListItems(itemsList) {
+    const key = this.props.getKey(itemsList[itemsList.length - 1]);
+    const expanded = this.state.expanded;
+    expanded.add(key);
+    this.setState({ expanded: expanded });
```
Comment (jasonLaster) [inline src/components/utils/ManagedTree.js]: same weirdness here. maybe we can do something where we look at the listItem props and then update the expanded set in batch?

**Comment by arthur801031** [comment]
Hi @jasonLaster 

I simplified the code for reveal in ManagedTree.js. Please check. Thank you very much for your help on this PR!! :)

File: src/components/utils/ManagedTree.js [hunk 10]
```diff
@@ -10,10 +10,22 @@ let ManagedTree = React.createClass({
...
     return {
       expanded: new Set(),
-      focusedItem: null
+      focusedItem: null,
+      treeListItems: []
     };
   },
 
+  componentWillReceiveProps(nextProps) {
+    this.setState({ treeListItems: nextProps.listItems });
```
Comment (jasonLaster) [inline src/components/utils/ManagedTree.js]: Hmm, what if we call expand list items from here. No state necessary.

**Comment by arthur801031** [comment]
Hi @jasonLaster ,

Great comments/suggestions! Please check. Thank you!

File: src/components/SourcesTree.js [hunk 11]
```diff
@@ -83,6 +92,12 @@ let SourcesTree = React.createClass({
     if (!nodeHasChildren(item)) {
       this.props.selectSource(item.contents.get("id"));
     }
+
+    // If listItems is not empty, focusedItem would be set incorrectly in
+    // ManagedTree.js's expandListItems function
+    if (this.state.listItems && this.state.listItems.length > 0) {
+      this.setState({ listItems: [] });
+    }
```
Comment (jasonLaster) [inline src/components/SourcesTree.js]: hmm, @arthur801031 this *also* seems strange.

I think it's here because `componentWillReceiveProps` does not check to see if `listItems` has changed. 
* perhaps check to see if listItems is different
* also could, `componentWillReceiveProps` be `componentDidUpdate` perhaps not, just curious.

File: src/components/utils/ManagedTree.js [hunk 12]
```diff
@@ -31,6 +38,16 @@ let ManagedTree = React.createClass({
     }
   },
 
+  expandListItems(treeListItems) {
+    const expanded = this.state.expanded;
+    for (let i = treeListItems.length - 1; i >= 0; i--) {
+      const key = this.props.getKey(treeListItems[i]);
+      expanded.add(key);
+    }
```
Comment (jasonLaster) [inline src/components/utils/ManagedTree.js]: you can clean this up by doing:

```js
treeListItems.forEach(item => this.state.expanded.add(this.props.getKey(item))
```

Not a big deal, but you know, we're perfectionists here :)

**Comment by arthur801031** [comment]
Hi @jasonLaster ,
Please check. Thank you!! :)

**Comment by jasonLaster** [comment]
Wow. At this rate we'll have a three line PR by Monday. 

Seriously though, will treat and likely never tomorrow morning.
